libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/mpp/TransformKotlinGranularMetadata.kt

```diff
@@ -15,22 +15,28 @@ import org.jetbrains.kotlin.gradle.plugin.sources.KotlinDependencyScope.*
 import org.jetbrains.kotlin.gradle.plugin.sources.getSourceSetHierarchy
 import org.jetbrains.kotlin.gradle.targets.metadata.ALL_COMPILE_METADATA_CONFIGURATION_NAME
 import org.jetbrains.kotlin.gradle.targets.metadata.KotlinMetadataTargetConfigurator
+import org.jetbrains.kotlin.gradle.utils.getValue
 import java.io.File
 import javax.inject.Inject
 
 open class TransformKotlinGranularMetadata
 @Inject constructor(
     @get:Internal
+    @field:Transient
     val kotlinSourceSet: KotlinSourceSet
 ) : DefaultTask() {
 
     @get:OutputDirectory
-    val outputsDir: File = project.buildDir.resolve("kotlinSourceSetMetadata/${kotlinSourceSet.name}")
+    val outputsDir: File by project.provider {
+        project.buildDir.resolve("kotlinSourceSetMetadata/${kotlinSourceSet.name}")
+    }
 
     @Suppress("unused") // Gradle input
     @get:InputFiles
     @get:PathSensitive(PathSensitivity.RELATIVE)
-    internal val allSourceSetsMetadataConfiguration = project.configurations.getByName(ALL_COMPILE_METADATA_CONFIGURATION_NAME)
+    internal val allSourceSetsMetadataConfiguration: FileCollection by lazy {
+        project.files(project.configurations.getByName(ALL_COMPILE_METADATA_CONFIGURATION_NAME))
+    }
```

Comment:
:x: `FileCollection` is already a "lazy" type, no need to wrap it inside a `by lazy {}`